### PR TITLE
React frontend to Sippy with new API's

### PR DESCRIPTION
**Pull request by stbenjam**
This is a React and [Material-UI](https://material-ui.com/) front-end to Sippy.   A live copy is being hosted 
http://sippy-stbenjam-dev.apps.sandbox-m2.ll9k.p1.openshiftapps.com/sippy-ng/.

There's still a bunch of things to do to have all of the current UI's features, but this can exist side-by-side with the current UI until it's feature complete. 

General documentation are in https://github.com/stbenjam/sippy/blob/new-ui/sippy-ng/README.md, and the new API's are documented in https://github.com/stbenjam/sippy/blob/new-ui/pkg/api/README.md.

All the older things work with one exception.  There's an existing react page at `/jobs`  that uses `/api/jobs`, I've re-written that to have this data created during testgrid conversion, and the react page has been moved to `sippy-ng/jobs/<release>/details`.

Sippy is also now built as a single binary, containing all the existing Static files and the new React material, it gets injected by a tool called go.rice. This simplifies the container building.

Todo's:
[ ] Canary test failures
[ ] Job runs with failure groups
[ ] Test impacting bugs
[ ] Test impacting components
[ ] Job impacting BZ components

### CODE REVIEW

**Comment by stbenjam** [comment]
For images to build, I think the builder will need to use ubi:8

**Comment by bparees** [comment]
> For images to build, I think the builder will need to use ubi:8

if you can switch to an image that's hosted on quay.io or registry.redhat.io that would be great.  Using docker.io images results in api throttling failures.  (The BuildConfig automation that builds+deploys sippy on app.ci actually replaces the FROM image with a different one to deal w/ this)

**Comment by stbenjam** [comment]
> > For images to build, I think the builder will need to use ubi:8
> 
> if you can switch to an image that's hosted on quay.io or registry.redhat.io that would be great. Using docker.io images results in api throttling failures. (The BuildConfig automation that builds+deploys sippy on app.ci actually replaces the FROM image with a different one to deal w/ this)

https://github.com/openshift/sippy/pull/254
https://github.com/openshift/release/pull/20957

**Comment by stbenjam** [comment]
It's building, and I think reached a pretty usable point.  What can I do to make this easier to review? I can break it up into a couple different PR's (e.g. new API's, and then the react parts) if it helps

File: Makefile [hunk 1]
```diff
@@ -1,10 +1,31 @@
...
+
+npm:
+	# For debugging
+	npm --version
+	# See https://github.com/facebook/create-react-app/issues/11174 about
+	# why we only audit production deps:
+	cd sippy-ng; [ -d node_modules ] || npm install --no-audit && npm audit --production
+
+clean:
+	rm -f sippy
```
Comment (stbenjam) [inline Makefile:31]: You're probably interested in the Makefile changes

File: pkg/sippyserver/server.go [hunk 2]
```diff
@@ -349,31 +371,111 @@ func (s *Server) variantsReport(w http.ResponseWriter, req *http.Request) {
...
+
+	// New API's
+	serveMux.HandleFunc("/api/health", s.jsonHealthReport)
+	serveMux.HandleFunc("/api/install", s.jsonInstallReport)
+	serveMux.HandleFunc("/api/jobs/details", s.jsonJobsDetailsReport)
+	serveMux.HandleFunc("/api/jobs", s.jsonJobsReport)
+	serveMux.HandleFunc("/api/releases", s.jsonReleasesReport)
+	serveMux.HandleFunc("/api/tests", s.jsonTestsReport)
+	serveMux.HandleFunc("/api/tests/details", s.jsonTestDetailsReport)
+	serveMux.HandleFunc("/api/upgrade", s.jsonUpgradeReport)
```
Comment (stbenjam) [inline pkg/sippyserver/server.go:478]: New API endpoint handler funcs

File: pkg/apis/api/types.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
+// Package api contains types suitable for use with Material UI data tables.
+package api
```
Comment (stbenjam) [inline pkg/apis/api/types.go:2]: This package has the new API types for tests and jobs

**Review by stbenjam — COMMENTED** [review]
@bparees I highlighted the changes I thought you'd care about to the core parts of Sippy

File: pkg/apis/sippyprocessing/v1/types.go [hunk 4]
```diff
@@ -136,18 +146,39 @@ type JobRunResult struct {
...
+	JobStatusTestFailure           JobStatus = "F"
+	JobStatusNoResults             JobStatus = "n"
+	JobStatusUnknown               JobStatus = "f"
+)
+
+type BuildResult struct {
+	Timestamp int       `json:"timestamp"`
+	Result    JobStatus `json:"result"`
+	URL       string    `json:"url"`
+}
```
[the remark this answers is not in the document]
Comment (bparees) [reply pkg/apis/sippyprocessing/v1/types.go:166]: this feels like it should be part of JobRunResult?

**Review by bparees — COMMENTED** [review]
one minor item to discuss

**Comment by stbenjam** [comment]
@bparees See 59abf271064d78f5271de929d0744b3bd6934c3a.  I had to remove an unused field called `HasUnknownFailures`, otherwise I was stuck in a catch 22 -- I had to get all the test results before I could create the JobRunResult, and link them to the JobResults.

**Pull request merged**